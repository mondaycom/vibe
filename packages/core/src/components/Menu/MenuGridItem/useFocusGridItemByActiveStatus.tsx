import type React from "react";
import { useEffect, useMemo } from "react";
import { useLastNavigationDirection } from "../Menu/hooks/useLastNavigationDirection";

export const useFocusGridItemByActiveStatus = ({
  wrapperRef,
  childRef,
  index,
  activeItemIndex,
  useDocumentEventListeners = false
}: {
  wrapperRef: React.RefObject<HTMLElement>;
  childRef: React.RefObject<HTMLElement>;
  index: number;
  activeItemIndex: number;
  useDocumentEventListeners?: boolean;
}) => {
  const { lastNavigationDirectionRef } = useLastNavigationDirection();
  const isActive = useMemo(() => index === activeItemIndex, [activeItemIndex, index]);

  useEffect(() => {
    if (useDocumentEventListeners) return;
    if (isActive) {
      childRef.current?.focus();
    } else {
      wrapperRef?.current?.blur?.();
    }
  }, [childRef, isActive, lastNavigationDirectionRef, wrapperRef, useDocumentEventListeners]);
};
