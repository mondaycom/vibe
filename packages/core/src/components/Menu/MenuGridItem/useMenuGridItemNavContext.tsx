import type React from "react";
import { useMemo } from "react";
import { NavDirections } from "../../../hooks/useFullKeyboardListeners";
import { useGridKeyboardNavigationContext } from "../../GridKeyboardNavigationContext/GridKeyboardNavigationContext";
import { type CloseMenuOption } from "../Menu/MenuConstants";
import { type GridElementRef } from "../../GridKeyboardNavigationContext/GridKeyboardNavigationContextConstants";

export const useMenuGridItemNavContext = ({
  wrapperRef,
  setActiveItemIndex,
  getPreviousSelectableIndex,
  getNextSelectableIndex,
  activeItemIndex,
  isUnderSubMenu,
  closeMenu
}: {
  wrapperRef?: GridElementRef;
  setActiveItemIndex?: (index: number) => void;
  getNextSelectableIndex?: (activeItemIndex: number) => number;
  getPreviousSelectableIndex?: (activeItemIndex: number) => number;
  activeItemIndex?: number;
  isUnderSubMenu?: boolean;
  closeMenu?: (option: CloseMenuOption) => void;
}) => {
  /*
   * This is an "adapter" between the Grid Keyboard Navigation mechanism and the Menu navigation mechanism.
   * Currently, the two mechanisms work a bit differently.
   * In the future, when the Menu component will use a GridKeyboardNavigationContext, this adapter shouldn't be needed anymore.
   */
  const innerKeyboardContext = useGridKeyboardNavigationContext([], wrapperRef);
  const keyboardContext = useMemo(
    () => ({
      onOutboundNavigation: (elementRef: React.RefObject<HTMLElement>, direction: NavDirections) => {
        innerKeyboardContext.onOutboundNavigation(elementRef, direction);

        switch (direction) {
          case NavDirections.UP:
            return setActiveItemIndex(getPreviousSelectableIndex(activeItemIndex));
          case NavDirections.DOWN:
            return setActiveItemIndex(getNextSelectableIndex(activeItemIndex));
          case NavDirections.LEFT:
            if (isUnderSubMenu) {
              closeMenu({});
            }
        }
      }
    }),
    [
      innerKeyboardContext,
      setActiveItemIndex,
      getPreviousSelectableIndex,
      activeItemIndex,
      getNextSelectableIndex,
      isUnderSubMenu,
      closeMenu
    ]
  );

  return keyboardContext;
};
