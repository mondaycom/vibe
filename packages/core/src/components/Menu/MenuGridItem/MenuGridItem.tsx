import React, { forwardRef, ReactElement, useCallback, useRef } from "react";
import useMergeRef from "../../../hooks/useMergeRef";
import { GridKeyboardNavigationContext } from "../../GridKeyboardNavigationContext/GridKeyboardNavigationContext";
import { useMenuGridItemNavContext } from "./useMenuGridItemNavContext";
import { useFocusGridItemByActiveStatus } from "./useFocusGridItemByActiveStatus";
import { useFocusWithin } from "../../../hooks/useFocusWithin";
import { VibeComponent, VibeComponentProps } from "../../../types";
import { CloseMenuOption } from "../Menu/MenuConstants";
import { ComponentDefaultTestId, getTestId } from "../../../tests/test-ids-utils";

export interface MenuGridItemProps extends VibeComponentProps {
  /**
   * The content of the menu grid item.
   */
  children?: ReactElement | ReactElement[];
  /**
   * If true, keyboard navigation will skip this item. This prop is also passed to the child.
   */
  disabled?: boolean;
  /**
   * A callback function to close the wrapping menu.
   */
  closeMenu?: (option: CloseMenuOption) => void;
  /**
   * The currently active index of the wrapping menu.
   */
  activeItemIndex?: number;
  /**
   * Callback function to set the active item index.
   */
  setActiveItemIndex?: (index: number) => void;
  /**
   * Function to get the next selectable index.
   */
  getNextSelectableIndex?: (activeItemIndex: number) => number;
  /**
   * Function to get the previous selectable index.
   */
  getPreviousSelectableIndex?: (activeItemIndex: number) => number;
  /**
   * The index of this menu grid item.
   */
  index?: number;
  /**
   * If true, this item is under a submenu instead of a top-level menu.
   */
  isUnderSubMenu?: boolean;
  /**
   * Callback function to open or close a submenu by its index.
   */
  setSubMenuIsOpenByIndex?: (index: number, isOpen: boolean) => void;
  /**
   * If true, event listeners will be attached to the document.
   */
  useDocumentEventListeners?: boolean;
}

const MenuGridItem: VibeComponent<MenuGridItemProps> & {
  isMenuChild?: boolean;
  isSelectable?: boolean;
} = forwardRef(
  (
    {
      className,
      id,
      children,
      index,
      activeItemIndex = -1,
      closeMenu,
      getNextSelectableIndex,
      getPreviousSelectableIndex,
      setActiveItemIndex,
      setSubMenuIsOpenByIndex,
      isUnderSubMenu = false,
      disabled = false,
      useDocumentEventListeners = false,
      "data-testid": dataTestId
    }: MenuGridItemProps,
    ref
  ) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRef(ref, componentRef);
    const childRef = useRef();

    const child = children && React.Children.only(children);
    if (!child) {
      console.error(
        "MenuGridItem can accept only a single element as first level child, this element is not valid: ",
        child
      );
    }

    const onFocusWithinChange = useCallback(
      (isFocusWithin: boolean) => {
        setSubMenuIsOpenByIndex(index, isFocusWithin);
        if (isFocusWithin) {
          setActiveItemIndex(index);
        }
      },
      [index, setActiveItemIndex, setSubMenuIsOpenByIndex]
    );
    const { focusWithinProps } = useFocusWithin({ onFocusWithinChange });

    useFocusGridItemByActiveStatus({
      wrapperRef: componentRef,
      childRef,
      activeItemIndex,
      index,
      useDocumentEventListeners
    });

    const keyboardContext = useMenuGridItemNavContext({
      wrapperRef: componentRef,
      setActiveItemIndex,
      getPreviousSelectableIndex,
      getNextSelectableIndex,
      activeItemIndex,
      isUnderSubMenu,
      closeMenu
    });

    return (
      <section
        ref={mergedRef}
        className={className}
        id={id}
        data-testid={dataTestId || getTestId(ComponentDefaultTestId.MENU_GRID_ITEM, id)}
        tabIndex={-1}
        role="grid"
        {...focusWithinProps}
      >
        <GridKeyboardNavigationContext.Provider value={keyboardContext}>
          {React.cloneElement(child, {
            ...child?.props,
            disabled,
            ref: childRef
          })}
        </GridKeyboardNavigationContext.Provider>
      </section>
    );
  }
);

Object.assign(MenuGridItem, {
  isMenuChild: true,
  isSelectable: true
});

export default MenuGridItem;
