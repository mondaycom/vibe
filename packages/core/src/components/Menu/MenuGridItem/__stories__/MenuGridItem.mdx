import { Canvas, Meta } from "@storybook/blocks";
import { TipOtherMenuItemComponents } from "./MenuGridItem.stories.helpers";
import * as MenuGridItemStories from "./MenuGridItem.stories";

<Meta of={MenuGridItemStories} />

# MenuGridItem

<code>MenuGridItem</code> can be used to place a grid-like, keyboard navigable container, inside a Menu. The user will be
able to interact and navigate into and from the grid in a natural way.

<Canvas of={MenuGridItemStories.Overview} />

## Props

Since <code>MenuGridItem</code> should be used only inside a <code>Menu</code>, almost all of the props below will be supplied automatically by the wrapping <code>Menu</code>.

### Import

```js
import { MenuGridItem } from "@vibe/core";
```

<PropsTable />

## Usage

<UsageGuidelines
  guidelines={[
    "MenuGridItem should always be used inside a Menu component",
    "A MenuGridItem can only have a single child",
    "The child of MenuGridItem should use the useGridKeyboardNavigation hook",
    "The child should use forwardRef, and have the same referenced element for useGridKeyboardNavigation.",
    <>
      Also, the referenced element should have a <code>tabIndex</code> value (probably -1).
    </>,
    <>
      MenuGridItem will pass the <code>disabled</code> prop to the child. The child should handle this prop and disable
      interactions.
    </>,
    <>
      To support a "disabled" mode, the child must have a prop named <code>disabled</code> (it will be automatically
      detected).
    </>,
    <>
      NOTE: Due to technical limitations, <code>useDocumentEventListeners</code> is not fully supported.
    </>
  ]}
/>

<TipOtherMenuItemComponents />

## Use cases and examples

### With disabled states

Disabled items will be "skipped" when using keyboard navigation. Try it for yourself!

<Canvas of={MenuGridItemStories.WithDisabledStates} />

### Inside sub-menus

Keyboard navigation is also supported in sub-menus

<Canvas of={MenuGridItemStories.InsideSubMenus} />

## Related components

<RelatedComponents componentsNames={["Menu", "MenuButton"]} />
