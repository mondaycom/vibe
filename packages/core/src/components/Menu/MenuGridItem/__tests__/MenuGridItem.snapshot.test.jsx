import { describe, it, expect } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import MenuGridItem from "../MenuGridItem";

const NO_OP = () => {};

describe("MenuGridItem renders correctly", () => {
  const FAKE_REQUIRED_PROPS = {
    closeMenu: NO_OP,
    setActiveItemIndex: NO_OP,
    getNextSelectableIndex: NO_OP,
    getPreviousSelectableIndex: NO_OP,
    setSubMenuIsOpenByIndex: NO_OP
  };

  it("with a child", () => {
    const tree = renderer
      .create(
        <MenuGridItem {...FAKE_REQUIRED_PROPS}>
          <div>Hello!</div>
        </MenuGridItem>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("with a disabled child", () => {
    // MenuGridItem should pass the disabled prop into DivWrapper
    const DivWrapper = ({ disabled }) => <div disabled={disabled}>this should be disabled</div>;
    const tree = renderer
      .create(
        <MenuGridItem disabled {...FAKE_REQUIRED_PROPS}>
          <DivWrapper />
        </MenuGridItem>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
