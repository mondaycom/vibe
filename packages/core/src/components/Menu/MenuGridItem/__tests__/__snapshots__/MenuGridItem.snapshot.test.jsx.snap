// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MenuGridItem renders correctly > with a child 1`] = `
<section
  data-testid="menu-grid-item"
  onBlur={[Function]}
  onFocus={[Function]}
  role="grid"
  tabIndex={-1}
>
  <div
    disabled={false}
  >
    Hello!
  </div>
</section>
`;

exports[`MenuGridItem renders correctly > with a disabled child 1`] = `
<section
  data-testid="menu-grid-item"
  onBlur={[Function]}
  onFocus={[Function]}
  role="grid"
  tabIndex={-1}
>
  <div
    disabled={true}
  >
    this should be disabled
  </div>
</section>
`;
